signing/signer.go

```diff
@@ -11,70 +12,90 @@ import (
 	}
 }
 
-// NewEdSignerFromKey builds a signer from a private key as byte buffer.
-func NewEdSignerFromKey(key PrivateKey, opts ...SignerOptionFunc) (*EdSigner, error) {
-	if len(key) != ed25519.PrivateKeySize {
-		log.Error("Could not create EdSigner from the provided buffer: buffer too small")
-		return nil, errors.New("buffer too small")
+// WithPrivateKey sets the private key used by EdSigner.
+func WithPrivateKey(priv PrivateKey) EdSignerOptionFunc {
+	return func(opt *edSignerOption) error {
+		if len(priv) != ed25519.PrivateKeySize {
+			return errors.New("could not create EdSigner from the provided key: too small")
+		}
+
+		keyPair := ed25519.NewKeyFromSeed(priv[:32])
+		if !bytes.Equal(keyPair[32:], priv.Public().(ed25519.PublicKey)) {
+			log.Error("Public key and private key do not match")
+			return errors.New("private and public do not match")
+		}
+
+		opt.priv = priv
+		return nil
 	}
+}
 
-	sgn := &EdSigner{priv: key}
-	keyPair := ed25519.NewKeyFromSeed(sgn.priv[:32])
-	if !bytes.Equal(keyPair[32:], sgn.priv.Public().(ed25519.PublicKey)) {
-		log.Error("Public key and private key do not match")
-		return nil, errors.New("private and public do not match")
-	}
-	for _, opt := range opts {
-		opt(sgn)
+// WithKeyFromRand sets the private key used by EdSigner using predictable randomness source.
+func WithKeyFromRand(rand io.Reader) EdSignerOptionFunc {
+	return func(opt *edSignerOption) error {
+		_, priv, err := ed25519.GenerateKey(rand)
+		if err != nil {
+			return fmt.Errorf("could not generate key pair: %w", err)
+		}
+
+		opt.priv = priv
+		return nil
 	}
-	return sgn, nil
 }
 
-// NewEdSignerFromRand generate signer using predictable randomness source.
-func NewEdSignerFromRand(rand io.Reader) *EdSigner {
-	_, priv, err := ed25519.GenerateKey(rand)
-	if err != nil {
-		log.Panic("Could not generate key pair err=%v", err)
-	}
-	return &EdSigner{priv: priv}
+// EdSigner represents an ED25519 signer.
+type EdSigner struct {
+	priv PrivateKey
+
+	prefix []byte
 }
 
 // NewEdSigner returns an auto-generated ed signer.
-func NewEdSigner(opts ...SignerOptionFunc) *EdSigner {
-	_, priv, err := ed25519.GenerateKey(nil)
-	if err != nil {
-		log.Panic("Could not generate key pair err=%v", err)
-	}
-	signer := &EdSigner{priv: priv}
+func NewEdSigner(opts ...EdSignerOptionFunc) (*EdSigner, error) {
+	cfg := &edSignerOption{}
+
 	for _, opt := range opts {
-		opt(signer)
+		if err := opt(cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	if cfg.priv == nil {
+		_, priv, err := ed25519.GenerateKey(nil)
+		if err != nil {
+			return nil, fmt.Errorf("could not generate key pair: %w", err)
+		}
+		cfg.priv = priv
 	}
-	return signer
+	sig := &EdSigner{
+		priv:   cfg.priv,
+		prefix: cfg.prefix,
+	}
+	return sig, nil
 }
 
 // Sign signs the provided message.
 func (es *EdSigner) Sign(m []byte) []byte {
-	if es.prefix != nil {
-		m = append(es.prefix, m...)
-	}
-	return ed25519.Sign2(es.priv, m)
+	msg := make([]byte, len(m)+len(es.prefix))
```

Comment:
check if prefix is nil?

Reply:
not needed, see my comment here: https://github.com/spacemeshos/go-spacemesh/pull/3877#discussion_r1056031159